source/openpulse/tests/test_openpulse_parser.py

```diff
@@ -329,6 +329,29 @@ def test_pragma_in_cal_block():
     assert _remove_spans(program) == expected
 
 
+def test_permissive_parsing(capsys):
+    p = """
+    cal {
+      int;
+    }
+    """
+
+    with pytest.raises(AttributeError, match=r"'NoneType' object has no attribute 'line'"):
+        # In this case, we do get an exception, but this is somewhat incidental --
+        # the antlr parser gives us a `None` value where we expect a `Statement`
+        parse(p, permissive=True)
```

Comment:
Isn't the point of `permissive=True` that parse will still return something? Might be nice to have a test case that shows the usefulness of `permissive=True` clearly.

Reply:
I don't have a good example where non-permissive parsing fails but permissive parsing succeeds. Prior to the recent grammar update we could have arranged for this with a `switch` statement within a `cal` block.

What `permissive=True` does is allow ANTLR to use a flexible error recovery mechanism (e.g. throw away tokens until it can resume parsing successfully), but the common situation right now is that the OpenPulse visitor chokes on the results of this. In the code example I added, what happens is the `OpenpulseStatementContext` has the `exception` field set to `NoViableAltException(None)` and the `stop` field set to `None`.  But the visitor code uses the `@span` decorator which does not check for the ANTLR exception, and so fails to correctly extract the span (tries to access `stop.line` but `stop` is `None`).

One thing we could do is file a more general issue and discuss whether we want to be able to recover in certain cases (which would require updating the visitor logic, at least here and possible elsewhere).

Reply:
This sounds like the permissive feature is not that usable in practice as it is?

If that's correct, I'm completely happy with merging this PR and us writing an issue describing the next step.

Reply:
Ok, I opened https://github.com/openqasm/openpulse-python/issues/30.